src/Tasks/Common/FileUtilities.MetadataReader.cs

```diff
@@ -21,31 +23,57 @@ namespace Microsoft.NET.Build.Tasks
 {
     static partial class FileUtilities
     {
+        private static Dictionary<string, (DateTime LastCheckedUtc, Version Version)> s_versionCache = new(StringComparer.OrdinalIgnoreCase /* Not strictly correct on *nix. Fix? */);
+
         private static Version GetAssemblyVersion(string sourcePath)
         {
-            using (var assemblyStream = new FileStream(sourcePath, FileMode.Open, FileAccess.Read, FileShare.Delete | FileShare.Read))
+            // TODO: is there a way to assert that it's a full path?
+
+            if (s_versionCache.TryGetValue(sourcePath, out var cacheEntry) 
+                && File.GetLastWriteTimeUtc(sourcePath) < cacheEntry.LastCheckedUtc)
```

Comment:
Why did you choose to use this inequality rather than checking for an exact last-write-time match on the file? If the user overwrites the file with another version and this version was created before the cache entry was added, the condition will hold and `GetAssemblyVersion` will be returning an incorrect version.

FWIW, analogous checks in RAR are based on exact timestamp checks: https://github.com/dotnet/msbuild/blob/ea1d6d99a376ac582216bba06b75f6bd9f3c7c64/src/Tasks/SystemState.cs#L359-L364

Reply:
Yeah, that's better.